### ISSUE TITLE
.KSplat Color issue with spherical harmonic

### ISSUE THREAD

**Issue by hyqshr** [mkkellogg/GaussianSplats3D#375]
While rendering my scene with `.ksplat` with `0` `sphericalHarmonicsDegree`, it worked well:

```
        const viewer = new GaussianSplats3D.DropInViewer({
            sharedMemoryForWorkers: false,
            showLoadingUI: true,
            format: GaussianSplats3D.SceneFormat.KSplat,
            sphericalHarmonicsDegree: 0
        });
```
<img width="789" alt="image" src="https://github.com/user-attachments/assets/ef5d256a-6547-45ee-9502-148bed8193ac">

But when I change the `sphericalHarmonicsDegree` to 1 or 2, the color seems to be broken: 

<img width="849" alt="image" src="https://github.com/user-attachments/assets/351c46a4-bcea-4aac-8c28-c5dc4ac874eb">

**Comment by mkkellogg** [comment]
Out of curiosity, how many splats are in your scene?

**Comment by hyqshr** [comment]
Only 1 splat. 

The broken color issue happens on `.ply` as well, but only when `sh_degree` is 1 (0 or 2 is fine). My model is trained from `splatfacto-w` (https://docs.nerf.studio/nerfology/methods/splatw.html) not sure if this the reason?

**Comment by mkkellogg** [comment]
Sorry let me rephrase, how many individual splat "blobs" are in your scene? If you press the 'I" key, you'll bring up an info panel that will give you that info:

![image](https://github.com/user-attachments/assets/c10895da-88e7-47a9-8d5f-1f61b55df62a)

**Comment by hyqshr** [comment]
Also - is there a python script to convert ply to ksplat? I think it can help a lot of folk's workflow. I can create a PR if needed

**Comment by mkkellogg** [comment]
If you want to send me your `.ksplat` or `.ply` file, I can tell you how many splats it has. My suspicion is that it has a large number, which can cause issues because there's not enough memory to store spherical harmonics data.

You could also load it in the online viewer here: https://projects.markkellogg.org/threejs/demo_gaussian_splats_3d.php and if you press 'I' in that viewer, it should bring up the info panel.

**Comment by hyqshr** [comment]
Below is the info panel. The weird thing is the broken color seems only happens at `sh_degree=1` for this `KSplat`. 0 and 2 are fine

<img width="925" alt="image" src="https://github.com/user-attachments/assets/c9d86889-7640-49a0-8b7a-e96fba5f2ed5">

**Comment by hyqshr** [comment]
I have another model have the same issue when `sh_degree` is 1 in viewer

<img width="783" alt="image" src="https://github.com/user-attachments/assets/f535f2f9-52c7-41bf-bd40-66bb28c3c082">

**Comment by mkkellogg** [comment]
That is pretty weird, would you mind sharing your `.ply` or `.ksplat` file and I can try to debug?

**Comment by hyqshr** [comment]
Thank you! Not sure why this link download `.ttf` format, but I can reproduce the error when I change it to `.ksplat`

https://forreal.blob.core.windows.net/forreal-model/0/splat-w-office.ksplat?sv=2023-01-03&st=2024-11-22T16%3A33%3A19Z&se=2024-11-23T16%3A33%3A19Z&sr=b&sp=r&sig=YVWvuRTBzwTA4UgOh%2BPyn5UQ9sKhU8EC01dmI4idyr4%3D

**Comment by mkkellogg** [comment]
Would you also be able to share your original `.ply` as well? And do you know how many SH degrees should be present in the file?

**Comment by PaulErpen** [comment]
Hi there,

I am experiencing a similar issue.
The colors are wrong when limiting the viewer to 1 spherical harmonics band. This happens given a model that has either been trained with the full range of spherical harmonics or a model that has been limited to 1 band.
It isn't specific to .ksplat files. I have this issue using .ply files, but it persists when converting to ksplat as well as supersplat-compressed ply files.

The following model has been trained using a maximum sh degree of 1:
[geo_gaussian-truck-low-1.ply](https://drive.google.com/file/d/1nRMPXbXEWnSda5HAE2S4HGH4vDSJ9P_w/view?usp=drive_link)

it looks fine when limiting the sh degree to 0:
![image](https://github.com/user-attachments/assets/3dc1a5c9-1093-4d10-a7b4-f215b8c321d4)

but falls apart when the spherical harmonics are set to 1:
![image](https://github.com/user-attachments/assets/fe6fe077-d870-4aaa-bd32-125abb120151)

Here are some stats about the model:
![image](https://github.com/user-attachments/assets/b0c77ea6-5469-4eed-9c53-aadbc9520add)

This also happens for models that are trained using all sh components:
[mcmc-vsc-truck-low-4_model(2).ply](https://drive.google.com/file/d/1i8QXDEwVjdLaTNaLIdCqrLnBW8rmbJ4Y/view?usp=sharing)

When limiting to sh 0:
![image](https://github.com/user-attachments/assets/50b15b61-ec8a-4648-8ac6-58f8fea6ef61)

When limiting to sh 1:
![image](https://github.com/user-attachments/assets/3706c121-bdab-4222-a426-f12d012027f7)

But its fine for all sh components (2 obviously):
![image](https://github.com/user-attachments/assets/0e7463e9-0590-48ed-b577-dbcdaed061f4)

Here are some stats about the other model:
![image](https://github.com/user-attachments/assets/2a23b479-a638-497d-893c-65fef307426c)

The supersplat editor seems to display the colors correctly:
![image](https://github.com/user-attachments/assets/8ef109c6-b646-44a2-845c-d679c918c10a)

I trained these models using different techniques. One is trained using geogaussian's original implementation and the other is trained using gsplat + mcmc.
Hope this is helpful :) It would be super appreciated if you find time to further look into this issue. If you need any assistance or further info, please let me know.
Thanks for making this awesome project open source.

**Comment by mkkellogg** [comment]
There is bug that currently manifests when you limit the spherical harmonics to degree 1, and it affects all file types. I have a fix in this branch: https://github.com/mkkellogg/GaussianSplats3D/tree/playcanvas_and_spz_compression

Hopefully I'll get a release out soon :)

**Comment by PaulErpen** [comment]
Hi thanks for the quick response :)
When checking out this branch and running the demo the issue persists:
![image](https://github.com/user-attachments/assets/7c2bb7f3-a340-4467-af6f-6c297cdd39dc)

maybe i am doing something wrong. I did the following:

```sh
git clone git@github.com:mkkellogg/GaussianSplats3D.git GaussianSplats3D-fixed
cd GaussianSplats3D-fixed
git fetch
git checkout remotes/origin/playcanvas_and_spz_compression
npm install
npm run build
npm run demo
```

Then I upload the file [mcmc-vsc-truck-low-4_model(2).ply](https://drive.google.com/file/d/1i8QXDEwVjdLaTNaLIdCqrLnBW8rmbJ4Y/view?usp=sharing)

**Comment by mkkellogg** [comment]
Sorry for the late response, it turns out I forgot to push some local changes to https://github.com/mkkellogg/GaussianSplats3D/tree/playcanvas_and_spz_compression. Want to give it another try?

**Comment by mkkellogg** [comment]
I totally forgot to update the online viewer, thanks for catching that! It should be fixed now.

**Comment by hyqshr** [comment]
Just tried again, still broken when sh=1

**Comment by mkkellogg** [comment]
Can you send my the file you're having issues with?

**Comment by hyqshr** [comment]
https://forreal.blob.core.windows.net/forreal-model/334/splat.ksplat?sv=2023-01-03&st=2025-02-07T14%3A46%3A30Z&se=2025-02-12T14%3A46%3A00Z&sr=b&sp=r&sig=dPSefgm9z5cHbpfFKulnRhareQJzEuOIQu9Xhf2%2BjFM%3D

**Comment by mkkellogg** [comment]
Are you doing any rotations or other transformations to the scene? I'm able to view it correctly with SH=1 using the latest in `main` (which should correspond to v0.4.7)
![Image](https://github.com/user-attachments/assets/022dcc49-bae7-45ce-8928-20d76ba53923)

**Comment by hyqshr** [comment]
I was just using the web viewer https://projects.markkellogg.org/threejs/demo_gaussian_splats_3d.php?art=1&cu=0,1,0&cp=0,1,0&cla=1,0,0&aa=false&2d=false&sh=1 with no rotations. Is this viewer most update-to-date?

**Comment by mkkellogg** [comment]
Interesting, the online viewer is up-to-date. In fact, I can follow the link you posted above, load `splat.ksplat`, and it all seems to work correctly. What OS and browser are you using?

**Comment by hyqshr** [comment]
macOS and chrome
